subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	/* Current LTV offset contains the offset in the AD structure.
+	 * The length and type are included in this offset.
+	 */
+	uint16_t current_ltv_offset;
+};
+
+static void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+static bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+static const struct bt_data *ad_stream_current_ltv_get(struct ad_stream *stream)
+{
+	const struct bt_data *current_ltv = &stream->ad[stream->ad_index].data[stream->data_index];
```

Comment:
This is confusing me. Why isn't it 
```suggestion
	const struct bt_data *current_ltv = stream->ad[stream->ad_index].data;
```
instead? `data_index` should not be relevant when retrieving an LTV structure. 

Furthermore, the naming should probably also become persistent. Either use `ad` or `ltv`. Using `ad_index` to get an `ltv` is odd, compared to using `ad_index` to get an `ad` or an `ltv_index` to get an `ltv`.

Reply:
It is because `stream->ad` is a two-dimensional array. `stream->ad` contains `stream->ad_len` elements of type `struct bt_ad`. Each `struct bt_ad` consists of `stream->ad[i].len` LTV groups (`struct bt_data`).

Reply:
Alright, I had misunderstood the dimensions then. 

I figured it was an array of `struct bt_ad` which in turn is a header and an array of data, but you are saying that we have an array of `struct bt_ad` which each has an array of `struct bt_data` which each has an array of data (alongside the header), effectively being a 3 dimensional array if you include the `data` array in `struct bt_data`. 

Can you elaborate on why it is two-dimensional array then? The API takes a single array of `struct bt_data`, so why are we adding an additional dimension? 

Another issue, which is course not introduce by this PR, is that we use `data` in `bt_ad` but also in `bt_data`, so it becomes unclear which `data` that `data_index` refers to, which is also causing some of the confusion I suppose.

Reply:
The `struct bt_ad` is used to add the complete local name to the advertising data without doing additional copying. See https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/adv.c#L650. 

In my opinion it would be more readable if we change then name of `struct bt_ad` to `struct bt_ad_list` or something similar.

Reply:
Perhaps `struct bt_data_list` (or `array` instead of `list`), as that is what is really is, right?

Reply:
Yes, something similar to that. In my opinion `data` is sometimes a bit too generic. But that is my personal opinion

Reply:
It is very generic :) I assume that `ad` also refers to advertising data, which is also very generic, and `bt_data` actually referring to LTV-structured data, which is slightly less generic. 

I think it is OK to keep the naming as is for now, as changing it would be significantly outside the scope of this PR, but we could add comments for the structs and fields to improve the understanding.

Reply:
Feel free to add suggestions for better comments to the structure members

Reply:
Added a few suggestions